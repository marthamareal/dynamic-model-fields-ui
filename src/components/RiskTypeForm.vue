<template>
  <b-modal id="modal-1" hide-footer title="Create Risk Type">
    <b-form @submit="onSubmit">
      <b-form-group label="Name:">
        <b-form-input
          v-model="form.name"
          type="text"
          required
          placeholder="Enter risk type name"
        ></b-form-input>
      </b-form-group>

      <b-form-group
      label="Description:"
      description="Include all information about the policy" >
        <b-form-textarea
          v-model="form.description"
          required
        ></b-form-textarea>
      </b-form-group>

      <label>Fields</label>

      <div class="card h-100">
      <div class="card-body">
        <b-form-group class = "form-inline" v-for="(field, index) in form.fields" :key="index">
          <b-form-input
            v-model="field.name"
            type="text"
            required
          ></b-form-input>
          <b-form-select
            id="input-3"
            v-model="field.field_type"
            :options="fieldTypes"
            required
          ></b-form-select>
          <span>
            <img v-if="index > 0" size="sm" class="cancel-img rounded-circle" @click="removeField(index)" src="../assets/cancel-icon.png"/>
            </span>
        </b-form-group>
        <div  @click="addField" >
        <img class="add-button" src="../assets/add-icon.png"/> Add another field
        </div>
      </div>
    </div>
    <b-button block variant="primary" class="mt-3 form" type="submit">Submit</b-button>
    </b-form>
  </b-modal>
</template>

<script>
import axios from 'axios'

export default {
  data () {
    return {
      form: {
        name: '',
        description: '',
        fields: [{name: 'date', field_type: 'date'}]
      },
      fieldTypes: [{ text: 'type', value: null }]
    }
  },
  async mounted () {
    await axios.get('https://dynamic-model-fields.herokuapp.com/api/v1/field-types/')
      .then(response => {
        let data = response.data
        this.fieldTypes = data.map(item => {
          return { text: item.name, value: item.id }
        })
      })
      .catch(response => {
        this.handleCreateResponses('Sory', 'Unable to fetch field types, Please contact our support for help.', 'danger')
      })
  },
  methods: {
    handleCreateResponses (title, message, variant) {
      this.$bvToast.toast(message, {
        title: title,
        toaster: 'b-toaster-top-center',
        variant: variant,
        solid: true
      })
    },

    onSubmit (evt) {
      evt.preventDefault()
      let formData = JSON.parse(JSON.stringify(this.form))

      axios.post('https://dynamic-model-fields.herokuapp.com/api/v1/risk-types/', formData)
        .then(response => {
          if (response.status === 201) {
            this.handleCreateResponses('Congrats', 'You have successfully added a new risk type', 'success')
          } else {
            this.handleCreateResponses('Sory', 'Unable to create risk type, something went wrong at the server.', 'danger')
          }
        })
        .catch(response => {
          this.handleCreateResponses('Oops!!', 'Something wrong has happened, Please contact our support for help.', 'danger')
        })
      document.querySelector('.close').click()
    },
    addField (e) {
      e.preventDefault()
      this.form.fields.push({ name: '', type: '' })
    },
    removeField (index) {
      this.form.fields.splice(index, 1)
    }
  }
}
</script>

<style>
.round-button{
  border-radius: 45%;
}
.cancel-img  {
height: inherit;
font-size: small;
padding-left: 20px;
}
.add-button {
  border: 1px solid #dfdfdf;
  padding: initial;
  height: 30px;
  width: 30px;
}
</style>
