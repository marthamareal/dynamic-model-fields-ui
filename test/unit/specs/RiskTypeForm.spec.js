import { mount } from '@vue/test-utils';
import RiskTypeForm from '@/components/RiskTypeForm'

jest.mock('axios', () => ({
    get: () => {
        const response = { 
            data: [
                {
                    "id": "37636f25-e6a5-4183-b932-c11decec5e61",
                    "name": "text"
                },
                {
                    "id": "633c77ab-e78d-4b27-b269-fbb6c598455b",
                    "name": "date"
                }
            ]
        }
        return new Promise((resolve, reject) => resolve(response))
    },
    // post: ({}) => {
    //     const response = { 
    //         data: {
    //             "id": "827c2adc-8d52-498b-a879-118187afdae5",
    //             "name": "Prize Policy2",
    //             "description": "This is my description feild which can not be empty",
    //             "fields": [
    //                 {
    //                     "id": "b33a255d-a3d3-40fd-88c2-af20e9a5d50b",
    //                     "name": "pname",
    //                     "field_type": "text",
    //                     "options": null
    //                 },
    //                 {
    //                     "id": "72a81077-e226-4996-890f-31ace670e9e9",
    //                     "name": "pnamse",
    //                     "field_type": "text",
    //                     "options": null
    //                 }
    //             ]
    //         }
    //     }
    //     return new Promise((resolve, reject) => resolve(response))
    // }
  }))

describe('RiskTypeForm.vue', () => {
    const wrapper = mount(RiskTypeForm)

    it('should render as expected', () => {
        wrapper.setData({
            fieldTypes: [{ text: 'type', value: null }],
            form: {
                name: 'name',
                description: 'sample',
                fields: [{name: 'date', field_type: 'date'}]
              },
        })
    expect(wrapper.html()).toMatchSnapshot();
  })

  it('should add new field form', () => {
        const addbutton = wrapper.findAll('.add-button')
        addbutton.trigger('click')
        const fieldSets = wrapper.findAll('.form-inline')
        expect(fieldSets.length).toEqual(2)
    })

    it('should remove a field form', () => {
        const fieldSets = wrapper.findAll('.form-inline')
        fieldSets.at(1).findAll('.cancel-img').trigger('click')
        expect(wrapper.findAll('.form-inline').length).toEqual(1)
    })

    it('should submit form', () => {
        wrapper.find("[type='submit']").trigger('submit')
        
    })
})
