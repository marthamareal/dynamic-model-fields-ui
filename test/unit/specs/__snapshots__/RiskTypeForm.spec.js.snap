// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RiskTypeForm.vue should render as expected 1`] = `
<b-modal id="modal-1" hide-footer="" title="Create Risk Type">
    <b-form>
        <b-form-group label="Name:">
            <b-form-input type="text" required="" placeholder="Enter risk type name"></b-form-input>
        </b-form-group>
        <b-form-group label="Description:" description="Include all information about the policy">
            <b-form-textarea required=""></b-form-textarea>
        </b-form-group> <label>Fields</label>
        <div class="card h-100">
            <div class="card-body">
                <b-form-group class="form-inline">
                    <b-form-input type="text" required=""></b-form-input>
                    <b-form-select id="input-3" options="[object Object]" required=""></b-form-select> <span><!----></span>
                </b-form-group>
                <div><img src="../assets/add-icon.png" class="add-button"> Add another field
                </div>
            </div>
        </div>
        <b-button block="" variant="primary" type="submit" class="mt-3 form">Submit</b-button>
    </b-form>
</b-modal>
`;
